### PR TITLE
daemon: refactor to loopback 4

### PR DESCRIPTION
**Pull request by FeelyChau**
Migrate to loopback 4:

- [x] plugin
- [x] pipeline
- [x] job
- [x] unit test

### CODE REVIEW

File: packages/daemon/package.json [hunk 1]
```diff
@@ -1,85 +1,79 @@
...
-    "url": ""
+    "type": "git"
   },
-  "author": "anonymous",
-  "license": "Apache 2.0",
-  "midway-server-options": {
-    "port": 6927,
-    "mode": "single"
+  "author": "FeelyChau <feely@outlook.com>",
+  "license": "",
```
Comment (yorkie) [inline packages/daemon/package.json:42]:
```suggestion
  "license": "Apache 2.0",
```

File: packages/daemon/package.json [hunk 2]
```diff
@@ -1,85 +1,79 @@
...
+    "@loopback/boot": "^3.1.0",
+    "@loopback/core": "^2.12.0",
+    "@loopback/repository": "^3.2.0",
+    "@loopback/rest": "^9.0.0",
+    "@loopback/rest-explorer": "^3.0.3",
+    "@loopback/service-proxy": "^3.0.3",
+    "@pipcook/costa": "^1.2.0",
+    "@pipcook/pipcook-core": "^1.2.0",
+    "debug": "^4.3.1",
+    "loopback-connector-sqlite3": "git+ssh://git@github.com:HugoPoi/loopback-connector-sqlite3.git",
```
Comment (yorkie) [inline packages/daemon/package.json:59]: It's good to specify the tag or commit in this URL

File: packages/daemon/public/index.html [hunk 3]
```diff
@@ -0,0 +1,88 @@
+<!DOCTYPE html>
+<html lang="en">
+
+<head>
+  <title>pipcook server</title>
```
Comment (yorkie) [inline packages/daemon/public/index.html:5]:
```suggestion
  <title>Pipcook Gateway</title>
```

File: packages/daemon/public/index.html [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+      a:hover, a:focus, a:active {
+        color: #2b78ff;
+      }
+    }
+  </style>
+</head>
+
+<body>
+  <div class="info">
+    <h1>daemon</h1>
```
Comment (yorkie) [inline packages/daemon/public/index.html:74]:
```suggestion
    <h1>Pipcook Gateway</h1>
```

File: packages/daemon/public/index.html [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    <p>Version 1.0.0</p>
```
Comment (yorkie) [inline packages/daemon/public/index.html:75]: Read the version from package.json?

File: packages/daemon/src/__tests__/README.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# Tests
+
+Please place your tests in this folder.
```
Comment (yorkie) [inline packages/daemon/src/__tests__/README.md:3]: It's good to describe the test directory here, just like this page: https://github.com/nodejs/node/tree/master/test#nodejs-core-tests

**Review by yorkie — CHANGES_REQUESTED** [review]

**Comment by yorkie** [comment]
Could this file be removed, too? https://github.com/alibaba/pipcook/blob/6d0f39f94cd94f4efcff67a56e71534e6d5cb98a/packages/daemon/bootstrap.js

**Comment by yorkie** [comment]
And this file is also required to be updated to new framework: https://github.com/alibaba/pipcook/blob/6d0f39f94cd94f4efcff67a56e71534e6d5cb98a/packages/daemon/benchmark/bootstrap.js

**Comment by yorkie** [comment]
One more question, is our migration scripts available?

**Comment by yorkie** [comment]
@FeelyChau We need to keep the name and upgrade the version to 2.0 at [package.json](https://github.com/alibaba/pipcook/blob/refactor/daemon/packages/daemon/package.json).

**Comment by FeelyChau** [comment]
> One more question, is our migration scripts available?

Not now, I'll commit it later.

**Comment by FeelyChau** [comment]
Pipeline service, pipeline model, job service, job model have been migrated, may I ask you to migrate the controllers and unit tests? @yorkie @rickycao-qy @WenheLI

**Comment by yorkie** [comment]
> Pipeline service, pipeline model, job service, job model have been migrated, may I ask you to migrate the controllers and unit tests? @yorkie @rickycao-qy @WenheLI

I think you could make detailed assignments, for example, specific tasks/functions list and its owner :)

**Comment by yorkie** [comment]
How about adding @SirM2z to the list?

**Comment by FeelyChau** [comment]
> How about adding @SirM2z to the list?

Of course, @SirM2z

**Comment by FeelyChau** [comment]
API list for pipeline controller:
* create: create pipeline from config
* list: list pipelines by filter
* remove: remove and uninstall all pipelines
* removeOne: remove and uninstall a pipeline by pipeline id
* getConfig: get the pipeline config by pipeline id
* get: get pipeline entity by pipeline id
* update:  update pipeline from config
* installById: start the installation process by pipeline id
* installByPipeline: install pipeline by pipeline id
* traceEvent: trace pipeline installation by trace id

@rickycao-qy May I ask you to migrate this controller and complete the unit tests, please?

**Comment by FeelyChau** [comment]
API list for job controller:
* run: run job from pipeline id
* list: list jobs by filter
* removeAll: remove all jobs
* remove: remove job by job id
* stop: cancel job by job id
* getParams: get the params by job id
* viewLog: get the logs by job id
* download: download the job output by job id
* get: get job entity by job id
* traceEvent: trace job running by trace id
@WenheLI May I ask you to migrate this controller, please?

**Comment by FeelyChau** [comment]
Change status to ready for CI, please **DO NOT** merge.

File: packages/cli/src/actions/board.ts [hunk 7]
```diff
@@ -1,5 +1,5 @@
 import { execAsync } from '../utils/common';
 
-export default async function() {
+export default async function() :Promise<void> {
```
Comment (yorkie) [inline packages/cli/src/actions/board.ts:3]:
```suggestion
export default async function(): Promise<void> {
```

File: packages/daemon/src/__tests__/unit/service/job.service.test.ts [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+test('should stop job without runnable', async (t) => {
+  const { jobService, jobRepository } = initJobService();
+  const mockFindById = jobRepository.stubs.findById.resolves({ id: 'mockId', status: 1 /* running */ } as any);
+  const mockUpdateById = jobRepository.stubs.updateById.resolves();
+  await jobService.stopJob('mockId');
+  t.true(mockUpdateById.calledOnceWith('mockId', { id: 'mockId', status: 4 /* canceled */ } as Job), 'mockSaveJob check');
+  t.true(mockFindById.calledOnceWith('mockId'), 'mockFindById check');
+});
+
+// test('start job and finish', async (t) => {
```
Comment (yorkie) [inline packages/daemon/src/__tests__/unit/service/job.service.test.ts:291]: test.skip

File: packages/daemon/src/__tests__/unit/utils/index.test.ts [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+import * as fs from 'fs-extra';
+import { join } from 'path';
+import * as request from 'request-promise';
+import * as ChildProcess from 'child_process';
+import { mockFunctionFromGetter } from '../../__helpers__';
+
+const result: any = {
+  // name: undefined,
+  dataCollect: '@pipcook/plugins-csv-data-collect',
+  dataCollectParams: { "url": "http://ai-sample.oss-cn-hangzhou.aliyuncs.com/image_classification/datasets/textClassification.zip" },
```
Comment (yorkie) [inline packages/daemon/src/__tests__/unit/utils/index.test.ts:14]:
```suggestion
  dataCollectParams: { 'url': 'http://ai-sample.oss-cn-hangzhou.aliyuncs.com/image_classification/datasets/textClassification.zip' },
```

**Review by yorkie — APPROVED** [review]

**Comment by yorkie** [comment]
Great works, rad man!

File: packages/daemon/src/__tests__/README.md (hunk 6 above)
Comment (FeelyChau) [reply packages/daemon/src/__tests__/README.md:3]: It's helpful for the contributors, I will supplement this document when refactoring the unit tests.

**Review by WenheLI — APPROVED** [review]
LGTM

**Review by rickycao-qy — APPROVED** [review]
LGTM

**Pull request merged**